runtime/src/message_processor.rs

```diff
@@ -66,38 +66,39 @@ fn verify_instruction(
 ) -> Result<(), InstructionError> {
     // Verify the transaction
 
-    // Make sure that program_id is still the same or this was just assigned by the system program,
-    //  but even the system program can't touch a credit-only account.
-    if pre.owner != post.owner && (!is_debitable || !system_program::check_id(&program_id)) {
+    // Make sure that program_id is still the same or this was just re-assigned by
+    //  the owner program
+    if pre.owner != post.owner && (!is_debitable || *program_id != pre.owner) {
         return Err(InstructionError::ModifiedProgramId);
     }
     // An account not assigned to the program cannot have its balance decrease.
-    if *program_id != post.owner && pre.lamports > post.lamports {
+    if *program_id != pre.owner && pre.lamports > post.lamports {
         return Err(InstructionError::ExternalAccountLamportSpend);
     }
     // The balance of credit-only accounts may only increase.
     if !is_debitable && pre.lamports > post.lamports {
         return Err(InstructionError::CreditOnlyLamportSpend);
     }
     // Only system accounts can change the size of the data.
-    if !system_program::check_id(&program_id) && pre.data.len() != post.data.len() {
+    if pre.data.len() != post.data.len() && !system_program::check_id(program_id) {
```

Comment:
@garious why do we have this restriction?

Reply:
@aeyakovenko wanted CreateAccount to be the only form of allocation. I think we were concerned about a resource attack since we only charge for a fixed number of BPF instructions - can't be expensive ones.

Reply:
ok, I'm not changing it in here, just changed the order of checks to put the speedier one first